client-runtime/src/main/java/com/microsoft/rest/v2/RestProxy.java

```diff
@@ -19,209 +20,169 @@
             }
-            if (method.isAnnotationPresent(POST.class)) {
-                info.method = "POST";
-                info.relativePath = method.getAnnotation(POST.class).value();
+            else if (method.isAnnotationPresent(POST.class)) {
+                methodProxyDetails.setMethodAndRelativePath("POST", method.getAnnotation(POST.class).value());
             }
-            if (method.isAnnotationPresent(PATCH.class)) {
-                info.method = "PATCH";
-                info.relativePath = method.getAnnotation(PATCH.class).value();
+            else if (method.isAnnotationPresent(PATCH.class)) {
+                methodProxyDetails.setMethodAndRelativePath("PATCH", method.getAnnotation(PATCH.class).value());
             }
-        }
-        return matrix;
-    }
 
-    private static final class MethodInfo {
-        private String method;
-        private String relativePath;
-        private Map<String, Integer> hostArgs;
-        private Map<String, Integer> pathArgs;
-        private Map<String, Integer> queryArgs;
-        private Map<String, Integer> headerArgs;
-        private Integer bodyArg;
-
-        private MethodInfo() {
-            hostArgs = new HashMap<>();
-            pathArgs = new HashMap<>();
-            queryArgs = new HashMap<>();
-            headerArgs = new HashMap<>();
+            final Annotation[][] allParametersAnnotations = method.getParameterAnnotations();
+            for (int parameterIndex = 0; parameterIndex < allParametersAnnotations.length; ++parameterIndex) {
+                final Annotation[] parameterAnnotations = method.getParameterAnnotations()[parameterIndex];
+                for (final Annotation annotation : parameterAnnotations) {
+                    final Class<? extends Annotation> annotationType = annotation.annotationType();
+                    if (annotationType.equals(HostParam.class)) {
+                        final HostParam hostParamAnnotation = (HostParam) annotation;
+                        methodProxyDetails.addHostSubstitution(hostParamAnnotation.value(), parameterIndex, !hostParamAnnotation.encoded());
+                    }
+                    else if (annotationType.equals(PathParam.class)) {
+                        final PathParam pathParamAnnotation = (PathParam) annotation;
+                        methodProxyDetails.addPathSubstitution(pathParamAnnotation.value(), parameterIndex, !pathParamAnnotation.encoded());
+                    }
+                    else if (annotationType.equals(QueryParam.class)) {
+                        final QueryParam queryParamAnnotation = (QueryParam) annotation;
+                        methodProxyDetails.addQuerySubstitution(queryParamAnnotation.value(), parameterIndex, !queryParamAnnotation.encoded());
+                    }
+                    else if (annotationType.equals(HeaderParam.class)) {
+                        final HeaderParam headerParamAnnotation = (HeaderParam) annotation;
```

Comment:
Yeah, I remembered that after I put out this pull request. I implemented the Headers annotation during my async work (the next pull request I'll send out), and rather than having one take precedence over the other, I made is so that they are additive with each others. It's allowed to have multiple values assigned to a single header, so if we have both HeaderParam and Headers annotations, then I just add them together (separate them by a ",").